File: 426-convert-binary-search-tree-to-sorted-doubly-linked-list/426-convert-binary-search-tree-to-sorted-doubly-linked-list.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val,Node _left,Node _right) {
        val = _val;
        left = _left;
        right = _right;
    }
};
*/

class Solution {
    Node head, tail;
    public Node treeToDoublyList(Node root) {
        if(root == null) return root;
        head = null;
        tail = null;
        dfs(root);
        head.left = tail;
        tail.right = head;
        return head;
    }
    public void dfs(Node root) {
        if(root == null) return ;
        dfs(root.left);
        if(head == null) head = root;
        else {
            tail.right = root;
            root.left =tail;
        }
        tail = root;
        dfs(root.right);
    }
}

File: 426-convert-binary-search-tree-to-sorted-doubly-linked-list/426-convert-binary-search-tree-to-sorted-doubly-linked-list.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
"""

class Solution:
    def treeToDoublyList(self, root: 'Optional[Node]') -> 'Optional[Node]':
        head, tail = None, None
        if not root: return root
        def inorder(root):
            nonlocal head,tail
            if not root:
                return
            inorder(root.left)
            if not head:
                head = root
            else:
                tail.right = root
                root.left = tail
            tail = root
            inorder(root.right)
        inorder(root)
        head.left = tail
        tail.right = head
        return head